### PR TITLE
Fix issue 22539 - slicing of returned ref scope static array should not be allowed

### PR DESCRIPTION
**Pull request by dkorpel**

### CODE REVIEW

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -7031,17 +7031,15 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
-                if (tf.isref && sc.func && !sc.intypeof && !(sc.flags & SCOPE.debug_)
-                    && tf.next.hasPointers() && sc.func.setUnsafe())
+                if (!checkAddressCall(sc, tf))
                 {
                     exp.error("cannot take address of `ref return` of `%s()` in `@safe` function `%s`",
                         ce.e1.toChars(), sc.func.toChars());
+                    exp.errorSupplemental("return type `%s` has pointers that may be `scope`", tf.next.toChars());
                 }
             }
         }
```
Comment (thewilsonator) [inline src/dmd/expressionsem.d:7045]: These

File: src/dmd/expressionsem.d [hunk 2]
```diff
@@ -7841,6 +7839,19 @@ private extern (C++) final class ExpressionSemanticVisitor : Visitor
...
+                    if (auto tf = ce.e1.type.isTypeFunction())
+                    {
+                        if (!checkAddressCall(sc, tf))
+                        {
+                            exp.error("cannot slice static array of `ref return` of `%s()` in `@safe` function `%s`",
+                                ce.e1.toChars(), sc.func.toChars());
+                            exp.errorSupplemental("return type `%s` has pointers that may be `scope`", tf.next.toChars());
+                        }
+                    }
+                }
```
Comment (thewilsonator) [inline src/dmd/expressionsem.d:7854]: are identical please put them in a function like you have for `checkAddressCall`

File: src/dmd/expressionsem.d [hunk 3]
```diff
@@ -12807,6 +12818,31 @@ bool checkAddressVar(Scope* sc, Expression exp, VarDeclaration v)
...
+ * that involves inspecting the function arguments and parameter types which
+ * is more complex.
+ *
+ * Params:
+ *      sc = context
+ *      tf = type of the function call in question
+ * Returns:
+ *      `true` if ok, `false` for error
+ */
+bool checkAddressCall(Scope* sc, TypeFunction tf)
```
Comment (thewilsonator) [inline src/dmd/expressionsem.d:12836]: is this private? if not it should be.

**Comment by WalterBright** [comment]
The problem was misidentified:

     https://issues.dlang.org/show_bug.cgi?id=22539#c2

This misidentification is also illustrated here:
```
exp.errorSupplemental("return type `%s` has pointers that may be `scope`", tf.next.toChars());
```
by use of "may". The compiler must know *for sure* if a pointer is scope or not.

**Comment by dkorpel** [comment]
> The problem was misidentified:

It wasn't, like I mentioned in the issue: we're dealing with static arrays here, not slices.

> This misidentification is also illustrated here:
> 
> ```
> exp.errorSupplemental("return type `%s` has pointers that may be `scope`", tf.next.toChars());
> ```
> 
> by use of "may". The compiler must know _for sure_ if a pointer is scope or not.

I added this in the DDoc comment of `checkAddressCall`

```
This is currently stricter than necessary: it can be safe to take the
address of a `ref` with pointer type when the pointer isn't `scope`, but
that involves inspecting the function arguments and parameter types which
is more complex.
 ```

It can actually be relaxed when the compiler checks that the function can't return scope pointers, but that adds more complexity so I leave that as a future enhancement.

File: src/dmd/expressionsem.d [hunk 4]
```diff
@@ -12807,6 +12818,31 @@ bool checkAddressVar(Scope* sc, Expression exp, VarDeclaration v)
...
 }
 
+/****************************************************
+ * Determine if the address of a `ref return` value of
+ * a function call with type `tf` can be taken safely.
+ *
+ * This is currently stricter than necessary: it can be safe to take the
+ * address of a `ref` with pointer type when the pointer isn't `scope`, but
+ * that involves inspecting the function arguments and parameter types which
+ * is more complex.
```
Comment (WalterBright) [inline src/dmd/expressionsem.d:12828]: This kind of inspection is just what escape.d is crafted to do. I'd prefer this be done.

**Comment by WalterBright** [comment]
> It wasn't, like I mentioned in the issue: we're dealing with static arrays here, not slices.

Yes, you are correct. My remaining issue is with the "may be scope". We have the technology to do it accurately, which is what escape.d is all about.

File: src/dmd/expressionsem.d (hunk 4 above)
Comment (dkorpel) [reply src/dmd/expressionsem.d:12828]: In that case I could use some help, because I don't know how to ask escape.d if a `CallExp` returns scope pointers.

**Comment by RazvanN7** [comment]
@WalterBright : @dkorpel mentioned that he would need some help to implement "how to ask escape.d if a CallExp returns scope pointers.". Do you have any pointers so that we can move this forward?

File: src/dmd/expressionsem.d (hunk 2 above)
Comment (dkorpel) [reply src/dmd/expressionsem.d]: I've factored more code out into `checkAddressCall` now

**Comment by dkorpel** [comment]
I suggest getting this in as an incremental improvement, because it prevents memory corruption in `@safe` code and it makes the behavior of taking the address with `&` consistent with static array slice `[]` on `ref` returns.

Relaxing it by proving that the returned value cannot be a `scope` pointer is also made more difficult by nested functions:
```D
void escape() @safe 
{
    int l;
    scope int* x = &l; // scope pointer
    ref int* f() {return x;}
    int** ptr = &f(); // not allowed
}
```
So I doubt that 'We have the technology to do it accurately".

**Review by RazvanN7 — APPROVED** [review]
@WalterBright This is definitely a step forward. I don't see any reason why we can't merge this and improve it in subsequent PRs

**Pull request merged**